File: java/2187. Minimum Time to Complete Trips/Solution.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        // binary search
        long start = 1;
        int maxTime = Arrays.stream(time)
            .max()
            .getAsInt();
        long end = (long) maxTime * totalTrips;

        while (start < end) {
            long mid = (start + end) / 2;
            long trips = totalTripsInTime(time, mid);

            if (trips < totalTrips) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return start;
    }

    public long totalTripsInTime(int[] time, long givenTime) {
        long trips = 0;
        for (int duration : time) {
            trips += givenTime / duration;
        }
        return trips;
    }
}

File: python/2187. Minimum Time to Complete Trips/solution.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        def totalTripsInGivenTime(times, givenTime):
            trips = 0
            for time in times:
                trips += (givenTime // time)
            return trips

        start = 1
        end = max(time) * totalTrips

        while (start < end) :
            mid = (end + start) // 2
            trips = totalTripsInGivenTime(time, mid)
            if trips < totalTrips:
                start = mid + 1
            else:
                end = mid

        return start
    